src/roiextractors/extractors/memmapextractors/memmapextractors.py

```diff
@@ -64,43 +62,80 @@ def get_num_channels(self):
+
+        num_frames = imaging.get_num_frames()
+        memmap_shape = imaging.video_structure.build_video_shape(n_frames=num_frames)
+        dtype = imaging.get_dtype()
+
+        # Load the memmap
+        video_memmap = np.memmap(
             save_path,
-            shape=(
-                imaging.get_num_frames(),
-                imaging.get_num_channels(),
-                imaging.get_image_size()[0],
-                imaging.get_image_size()[1],
-            ),
-            dtype=imaging.get_dtype(),
+            shape=memmap_shape,
+            dtype=dtype,
             mode="w+",
         )
 
-        if verbose:
-            for ch in range(imaging.get_num_channels()):
-                print(f"Saving channel {ch}")
-                for i in tqdm(range(imaging.get_num_frames())):
-                    plane = imaging.get_frames(i, channel=ch)
-                    video_to_save[ch, i] = plane
-        else:
-            for ch in range(imaging.get_num_channels()):
-                for i in range(imaging.get_num_frames()):
-                    plane = imaging.get_frames(i, channel=ch)
-                    video_to_save[ch, i] = plane
+        if file_size_in_bytes < buffer_size_in_bytes:
+            video_data_to_save = imaging.get_frames()
+            video_memmap[:] = video_data_to_save
 
-        video_to_save.flush()
+        else:
+            chunk_size_in_bytes = int(buffer_size_in_bytes)
+            type_size = np.dtype(dtype).itemsize
+
+            n_channels = imaging.get_num_channels()
+            pixels_per_frame = n_channels * np.product(imaging.get_image_size())
+            bytes_per_frame = type_size * pixels_per_frame
+            frames_per_chunk = chunk_size_in_bytes // bytes_per_frame
+
+            # If the chunk size is smaller than one image, the iterator goes over one image only.
+            frames_per_chunk = max(frames_per_chunk, 1)
+            iterator = range(0, num_frames, frames_per_chunk)
+            if verbose:
+                iterator = tqdm(iterator, ascii=True, desc="Writing to .dat file")
+
+            for frame in iterator:
+                end_frame = min(frame + frames_per_chunk, num_frames)
+
+                # Get the video chunk
+                video_chunk = imaging.get_video(start_frame=frame, end_frame=end_frame)
+
+                # Fit the video chunk in the memmap array
+                indices = np.arange(start=frame, stop=end_frame)
+                axis_to_expand = (
+                    imaging.video_structure.rows_axis,
+                    imaging.video_structure.columns_axis,
+                    imaging.video_structure.num_channels_axis,
+                )
+                indices = np.expand_dims(indices, axis=axis_to_expand)
+                frame_axis = imaging.video_structure.frame_axis
+                np.put_along_axis(arr=video_memmap, indices=indices, values=video_chunk, axis=frame_axis)
```

Comment:
Confused why you still need all this axis management here; if we enforce the `(frames, rows, cols, channels)` structure, and enforce that calls to `get_video` return the same data in that similar convention, then you should just be able to set the `video_memmap` values directly as

```
video_memmap[n_frames_in_buffer, ...] = video_buffer  # 'video_chunk' -> 'video_buffer' as well
```

Reply:
This is done in the #125.

Reply:
OK